### ISSUE TITLE
Speed up creating SQL meta db

### ISSUE THREAD

**Issue by dkirkby** [dkirkby/bossdata#2]
The time to create the db is roughly the same (~25 minutes) for the lite (~20 columns) and full (~1K columns), so it looks like the bottleneck is some constant overhead of the INSERT statement. Try batching the INSERTs using the sqlite executemany() command, to see if this helps.

**Comment by dkirkby** [comment]
Can also set the `iter_row_buffer` parameter when opening the FITS version of spAll to buffer on the input side.

**Comment by dcunning11235** [comment]
For the "lite" version, I think this is an issue with the Astropy read; that is where this is stalling for me.  I also see > 4.5 GB of memory in use before my laptop's OS kills the process.

From reading https://github.com/astropy/astropy/issues/3334, I think changing the format parameter to e.g. ascii.fixed_width (or something similar) may be a partial solution.

After that, breaking SQL inserts into larger transactions should help a lot (without implicit commits after each insert, which I believe is happening now.)

Given data files of ~1 GB (already), breaking the read-parse-SQL-insert into chunks may be a good path forward.

I'm going to start working on this, see where I get.

**Comment by dkirkby** [comment]
@dcunning11235 Please go ahead and post updates here.  I can show you how to submit a pull request if you get something working.

I guess you are getting stuck on these lines:

```
    with gzip.open(sp_all_path, mode='r') as f:
        table = astropy.table.Table.read(f, format='ascii')
```

Since the gzipped "lite" spAll is only 115Mb (and it uncompresses to 757Mb), I was assuming that there wouldn't be any problem parsing it and building the `astropy.table.Table` in memory.  I'm surprised that you saw >4.5Gb usage though.  We might need to use the (presumably more efficient) [`numpy.loadtxt`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.loadtxt.html) instead.

**Comment by dcunning11235** [comment]
I've solved my issue with out of memory by:
1.)  chunking the parse and insertion
2.)  specifying the format as `ascii.fixed_width_two_line` and setting `guess=False` in the call, which supposedly (I've not verified) helps with memory usage.

``` python
def create_meta_lite(sp_all_path, db_path, verbose=True):
    chunk_size = 50000
    init_position = 2
    position=init_position

    with gzip.open(sp_all_path, mode='r') as f:
        table = astropy.table.Table.read(f, data_start=position,
                                         data_end=position+chunk_size,
                                         format='ascii.fixed_width_two_line',
                                         guess=False)

```

...[create table based on metadata form initial read/chunk]...

``` python
        sql = 'INSERT INTO meta VALUES ({values})'.format(values=','.join('?' * num_cols))
        while len(table) > 0:
            cursor.executemany(sql, table)

            if verbose:
                print(position+chunk_size-init_position, "rows...")

            position += chunk_size
            table = astropy.table.Table.read(f, data_start=position,
                                         data_end=position+chunk_size,
                                         format='ascii.fixed_width_two_line',
                                         guess=False)
```

Having done that, I'm now struggling with trying improve performance:  Running this from spyder, in a VM (aka, not a higher performance environment), I'm getting about 70s to parse and insert 50k rows, with roughly a split of 12s of parsing and 58s of inserting.  Focusing on the SQLite side of things, I've --at best-- shaved maybe 1 or 2 seconds off having tried turning off logging/rollbacks, moving to using `cursor.executemany`  rather than `cursor.execute`, creating a unique index after the table is filled rather than a primary key before hand... basically all for nada, which is somewhat curious.

But I'm feeling this isn't a mystery that needs to be solved right now.  This may be as good as it gets (~1.4 ms / row, parse and insert, on my VM), and this is a one time (or maybe two time) process.

The chunking approach is necessary, however, at least on some machines.  I'll fork and include this new code, but I am concerned that doing the above will be much slower for those with machines happily run the original code: but I can't compare because I can't run the original code to completion on either of my machines!

**Comment by dkirkby** [comment]
Having something that works on all platforms is more important than having it run fast on a few platforms, especially in this case where the code is only run infrequently.

**Comment by dkirkby** [comment]
This has been fixed with #32 and #33
